### DISCUSSION TITLE
Scenario where DesignMode is false from the NET WinForms OOP designer

### DISCUSSION THREAD

**Discussion by tracktownsoftware** [dotnet/winforms#7448] [Q&A]
### Description

When using my NET6 Windows Forms custom control to display a dialog at design time, the DesignMode value is incorrectly false even though it is running in the DesignToolsServer process.

### Reproduction Steps

My github sample to reproduce:
https://github.com/tracktownsoftware/MicrosoftWinFormsDesignerSDK_DesignModeBug

The above repository includes a Nuget package you can use to test and reproduce. The nuget package contains a custom NET6 WinForms "MyButton" control with WinForms Designer Extensibility SDK design-time support.

### Expected behavior

DesignMode should be true because it is running in the DesignToolsServer process.

### Actual behavior

DesignMode is false.

See https://github.com/tracktownsoftware/MicrosoftWinFormsDesignerSDK_DesignModeBug/blob/main/README.md

### Known Workarounds

Rather than using DesignMode check for System.Diagnostics.Process.GetCurrentProcess().ProcessName=="DesignToolsServer"

**Comment by RussKie** [comment]
`DesignMode` is only indicating whether the control is being designed, which is quite specific activity. If a control is on a form or a user control that is being designed then your control isn't being designed but is being run. This is a common source of confusion for many developers. There were long discussions about the subject in https://github.com/dotnet/winforms/issues/3300 and https://github.com/dotnet/winforms/issues/4146.
So when you're invoking "_Dialog scenario showing DesignMode incorrectly false at design-time..._" menu you're no longer _designing_ the control but rather _executing_ it. In this case, even the new [`IsAncestorSiteInDesignMode` API](https://docs.microsoft.com/dotnet/api/system.windows.forms.control.isancestorsiteindesignmode) won't be of any help.

**Reply by RussKie** [reply]
Also it is worth noting, that the designtoolserver process must not show any [modal UI](https://github.com/tracktownsoftware/MicrosoftWinFormsDesignerSDK_DesignModeBug/blob/05615f7a08b2418702a9f093ee173f97a8196fc4/MyButtonLibrary2.Design/MyButtonActionList.cs#L40-L46) (or any UI in general) as it could lock the VS process. The UI must be invoked on the VS side through the notification mechanisms (i.e., the server sends notifications to the client to show the UI, collect user input and send the data back to the server).

**Reply by tracktownsoftware** [reply]
NET6 WPF DesignerProperties.GetIsInDesignMode(this) does return true in this scenario, so I was surprised NET6 WinForms returns false.

I understand the modal UI issue, but for us to implement it "properly" on the VS side my understanding is we will have to port our complicated NET6 component and its UI to .NET Framework. Is there a way to leverage our NET6 component in a VS side editor at design-time that maybe I missed? Thanks!

**Reply by RussKie** [reply]
Sadly, components will need to be re-architected and re-designed to work in the out-of-process designer. And this is not to just accommodate .NET (although that largly necessitated the move of the designer out of proc), but for .NET Framework as well, if components target or must be available on x86 platform, becuase VS2022 is now a 64-bit process.
@KlausLoeffelmann has written a blog post explaining the reasons and the associated challenges, you can find it here: https://devblogs.microsoft.com/dotnet/state-of-the-windows-forms-designer-for-net-applications/.

> my understanding is we will have to port our complicated NET6 component and its UI to .NET Framework.

I am unsure if this is entirely correct - that is, yes, the code will need to be split into the client and the server parts. The server part would then be responsible for running the .NET code and rendering the control/component. The client part is run inside the VS on net472, and it responsible for showing dialogs and the property browser interactions. The data for the dialogs and the property browser should be coming from the server process (via request/response or notification channels). We are aware of the large amount of biolerplate code that needs to be written to facilitate that, and we're investigating ways to reduce/remove the burden.

Unfortunately the control authoring documentation that we promised few month ago is still no ready, and we appreciate that it makes things very difficult for anyone developing controls. There's another related discussion here: https://github.com/dotnet/winforms/discussions/7073. @KlausLoeffelmann has sample repo that demonstrates the use of the Designer Extensibility SDK: https://github.com/dotnet/winforms/discussions/7073#discussioncomment-2618155.
